File: services/semantic/src/services/work_identifier.py
from __future__ import annotations

import re


OPENALEX_WORK_RE = re.compile(r"^W\d+$", re.IGNORECASE)
OPENALEX_URL_RE = re.compile(r"^https?://(?:www\.|api\.)?openalex\.org/(?:works/)?", re.IGNORECASE)
DOI_RE = re.compile(r"10\.\d{4,9}/\S+", re.IGNORECASE)


def normalize_openalex_work_id(value: str | None) -> str | None:
    text = _clean(value)
    if not text:
        return None
    text = OPENALEX_URL_RE.sub("", text).strip().strip("/")
    if OPENALEX_WORK_RE.match(text):
        return text.upper()
    return None


def openalex_aliases(value: str | None) -> list[str]:
    normalized = normalize_openalex_work_id(value)
    if not normalized:
        return []
    return [normalized, f"https://openalex.org/{normalized}"]


def normalize_doi(value: str | None) -> str | None:
    text = _clean(value)
    if not text:
        return None
    text = re.sub(r"^https?://(?:dx\.)?doi\.org/", "", text, flags=re.IGNORECASE)
    text = re.sub(r"^(?:doi\s*:|doi\s+)", "", text, flags=re.IGNORECASE).strip()
    match = DOI_RE.search(text)
    if not match:
        return None
    return match.group(0).strip().strip(".,;").lower()


def doi_aliases(value: str | None) -> list[str]:
    normalized = normalize_doi(value)
    if not normalized:
        return []
    return [
        normalized,
        f"https://doi.org/{normalized}",
        f"http://doi.org/{normalized}",
        f"https://dx.doi.org/{normalized}",
        f"http://dx.doi.org/{normalized}",
    ]


def normalize_work_identifier(value: str | None) -> str | None:
    return normalize_openalex_work_id(value) or normalize_doi(value) or _clean(value)


def work_identifier_aliases(value: str | None) -> tuple[str, list[str]]:
    openalex = normalize_openalex_work_id(value)
    if openalex:
        return "openalex", openalex_aliases(openalex)
    doi = normalize_doi(value)
    if doi:
        return "doi", doi_aliases(doi)
    cleaned = _clean(value)
    return "submission", [cleaned] if cleaned else []


def _clean(value: str | None) -> str | None:
    if value is None:
        return None
    text = str(value).strip()
    return text or None


__all__ = [
    "doi_aliases",
    "normalize_doi",
    "normalize_openalex_work_id",
    "normalize_work_identifier",
    "openalex_aliases",
    "work_identifier_aliases",
]

File: apps/frontend/src/utils/workIdentifier.ts
const OPENALEX_WORK_RE = /^W\d+$/i
const OPENALEX_URL_RE = /^https?:\/\/(?:www\.|api\.)?openalex\.org\/(?:works\/)?/i
const DOI_RE = /10\.\d{4,9}\/\S+/i

function clean(value: string | undefined | null) {
  const text = String(value ?? '').trim()
  return text || ''
}

export function normalizeOpenAlexWorkId(value: string | undefined | null) {
  const text = clean(value).replace(OPENALEX_URL_RE, '').replace(/^\/+|\/+$/g, '')
  return OPENALEX_WORK_RE.test(text) ? text.toUpperCase() : ''
}

export function normalizeDoi(value: string | undefined | null) {
  let text = clean(value)
  text = text.replace(/^https?:\/\/(?:dx\.)?doi\.org\//i, '')
  text = text.replace(/^doi\s*:?\s*/i, '')
  const match = text.match(DOI_RE)
  return match ? match[0].replace(/[.,;]+$/g, '').toLowerCase() : ''
}

export function normalizeWorkIdentifier(value: string | undefined | null) {
  return normalizeOpenAlexWorkId(value) || normalizeDoi(value) || clean(value)
}

export function normalizeWorkIdentifierList(values: Array<{ id?: string }> | string[] | undefined) {
  const seen = new Set<string>()
  const result: string[] = []
  for (const item of values ?? []) {
    const raw = typeof item === 'string' ? item : item.id
    const normalized = normalizeWorkIdentifier(raw)
    if (!normalized || seen.has(normalized)) continue
    seen.add(normalized)
    result.push(normalized)
  }
  return result
}
